direct-auth-samples/src/main/java/com/okta/spring/example/controllers/LoginController.java

```diff
@@ -0,0 +1,303 @@
+import com.okta.idx.sdk.api.model.AuthenticationStatus;
+import com.okta.idx.sdk.api.model.AuthenticatorType;
+import com.okta.idx.sdk.api.model.AuthenticatorUIOption;
+import com.okta.idx.sdk.api.model.ChangePasswordOptions;
+import com.okta.idx.sdk.api.model.IDXClientContext;
+import com.okta.idx.sdk.api.model.RecoverPasswordOptions;
+import com.okta.idx.sdk.api.model.UserProfile;
+import com.okta.idx.sdk.api.model.VerifyAuthenticatorOptions;
+import com.okta.idx.sdk.api.response.AuthenticationResponse;
+import com.okta.idx.sdk.api.response.NewUserRegistrationResponse;
+import com.okta.idx.sdk.api.response.TokenResponse;
+import com.okta.idx.sdk.api.wrapper.AuthenticationWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.servlet.ModelAndView;
+
+import javax.servlet.http.HttpSession;
+import java.util.List;
+
+@Controller
+public class LoginController {
+
+    private final Logger logger = LoggerFactory.getLogger(LoginController.class);
+
+    @Autowired
+    private IDXClient client;
+
+    @PostMapping("/custom-login")
+    public ModelAndView handleLogin(@RequestParam("username") String username,
+                                    @RequestParam("password") String password,
+                                    HttpSession session) {
+
+        TokenResponse tokenResponse =
+                (TokenResponse) session.getAttribute("tokenResponse");
+
+        // render existing token response if a successful one is already present in session
+        if (tokenResponse != null) {
+            ModelAndView mav = new ModelAndView("home");
+            AuthenticationResponse authenticationResponse = new AuthenticationResponse();
+            authenticationResponse.setAuthenticationStatus(AuthenticationStatus.SUCCESS);
+            authenticationResponse.setTokenResponse(tokenResponse);
+            mav.addObject("authenticationResponse", authenticationResponse);
+            return mav;
+        }
+
+        // trigger authentication
+        AuthenticationResponse authenticationResponse =
+                AuthenticationWrapper.authenticate(client, new AuthenticationOptions(username, password));
+
+        // populate login view with errors
+        if (authenticationResponse.getAuthenticationStatus() != AuthenticationStatus.SUCCESS) {
+            ModelAndView mav = new ModelAndView("login");
+            mav.addObject("errors", authenticationResponse.getErrors());
+            return mav;
+        }
+
+        // success
+        ModelAndView mav = new ModelAndView("home");
+        mav.addObject("authenticationResponse", authenticationResponse);
+
+        // store attributes in session
+        session.setAttribute("user", username);
```

Comment:
Looks like this is only used in the view and could be passed to it rather than fetched from session.

Reply:
updated.